accepted/00053-disallow-downloading-when-patching.md

```diff
@@ -0,0 +1,61 @@
+
+# Detailed design
+- two new functions are implemented in [zypper's execution module](https://github.com/openSUSE/salt/blob/openSUSE-2018.3.0/salt/modules/zypper.py). Given a list of packages or patches, return `True` if any downloading is required
+
+Those functions can leverage `zypper`'s estimation of downloaded bytes in dry-run mode to determine if any downloading is needed:
+
+```
+minion:~ # zypper --non-interactive --xmlout install --dry-run zsh
+<?xml version='1.0'?>
+<stream>
+
+...
+
+<install-summary download-size="0" space-usage-diff="7389792" packages-to-change="1">
+
+...
+</stream>
+```
+
+Those functions should accept a subset of parameters from the respective original state modules ([pkg.installed](https://docs.saltstack.com/en/2018.3/ref/states/all/salt.states.pkg.html#salt.states.pkg.installed) and [pkg.patch_installed](https://docs.saltstack.com/en/2018.3/ref/states/all/salt.states.pkg.html#salt.states.pkg.patch_installed)).
+
+
+- The Server's [package installation](https://github.com/uyuni-project/uyuni/blob/c8ffe6b9425392f5235864ad070646bb8ebc2ecb/susemanager-utils/susemanager-sls/salt/packages/pkginstall.sls) and [patch application](https://github.com/uyuni-project/uyuni/blob/c8ffe6b9425392f5235864ad070646bb8ebc2ecb/susemanager-utils/susemanager-sls/salt/packages/patchinstall.sls) modules will call those functions and conditionally skip actual installation/application via `unless` clauses
+
+- This should only happen if the minion's `mgr_fail_installation_if_download_is_required` Pillar is set
+
+- Pillars controlling this feature can be set at a global, per-minion or per System Group basis ([similarly to what we do to set the package download endpoint override](https://github.com/SUSE/doc-susemanager/pull/366/files#diff-8bf787463eb899039a59878b8b2ce800R171)).
+
+
+# Drawbacks
+ - this is limited to Salt minions. Traditional clients have no such feature
+ - only covers Zypper at this point. Yum/DNF can be implemented by leveraging the [--cacheonly](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/deployment_guide/sec-working_with_yum_cache#idm140707649391200) option
+ - no UI
+
+# Alternatives
+- use Action Chains
+  - pro: more visible to users
+  - con: could have scalability issues, more difficult to implement
+- implement custom modules instead of adding new functions to Salt's core
+  - pro: does not need a pull request against Salt
+  - con: the functionality might be useful in other contexts
+- add this feature directly to Zypper via a commandline switch (analoguous to [--cacheonly](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/deployment_guide/sec-working_with_yum_cache#idm140707649391200) in `yum` and `dnf`)
+  - pro: arguably cleaner design, does not need two calls to Zypper
+  - con: updated Zypper would be needed, more difficult to implement
```

Comment:
I see this alternative as a goal for the future - as a second step. It has some more "pros" then listed.
In case there is a reposync between the download-only and the install/update action, it might happen that a newer package is available. This would cause a failed action as zypper would try to download the newest package.

With something like `--cacheonly`, zypper would just calculate based on the cache and install what it has cached.
It would also prevent a download of metadata, which could slow down the execution time.

But as we might need this feature also for very old clients (SLE11) where we will not get a zypper update anymore, we should implement this RFC as it is now as a first step.

But we should open a Feature request for `zypper` to implement `--cacheonly` and as soon as some distributions have it, we should switch the implementation to this alternative.